src/cose/mac0.rs

```diff
@@ -0,0 +1,466 @@
+/// let unprotected = coset::HeaderBuilder::new().key_id(b"11".to_vec()).build();
+/// let builder = coset::CoseMac0Builder::new()
+///     .protected(protected)
+///     .unprotected(unprotected)
+///     .payload(b"This is the content.".to_vec());
+/// let prepared = PreparedCoseMac0::new(builder, None, None, true).unwrap();
+/// let signature_payload = prepared.signature_payload();
+/// let signature = tag(signature_payload, &signer).unwrap();
+/// let cose_mac0 = prepared.finalize(signature);
+/// fn tag(signature_payload: &[u8], s: &Hmac<Sha256>) -> anyhow::Result<Vec<u8>> {
+///     let mut mac = s.clone();
+///     mac.reset();
+///     mac.update(signature_payload);
+///     Ok(mac.finalize().into_bytes().to_vec())
+///  }
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct PreparedCoseMac0 {
+    tagged: bool,
+    cose_mac0: CoseMac0,
+    tag_payload: Vec<u8>,
+}
+
+#[derive(Clone, Debug)]
+pub struct CoseMac0 {
+    tagged: bool,
+    pub inner: coset::CoseMac0,
+}
+
+/// Errors that can occur when building, signing or verifying a COSE_Mac0.
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("the COSE_Mac0 has an attached payload but an detached payload was provided")]
+    DoublePayload,
+    #[error("the COSE_Mac0 has a detached payload which was not provided")]
+    NoPayload,
+    #[error("tag did not match the structure expected by the verifier: {0}")]
+    MalformedTag(MacError),
+    #[error("tag is already present")]
+    AlreadyTagged,
+    #[error("error occurred when tagging COSE_Mac0: {0}")]
+    Tagging(MacError),
+    #[error("unable to set ClaimsSet: {0}")]
+    UnableToDeserializeIntoClaimsSet(CoseError),
+}
+
+/// Result with error type: [`Error`].
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+/// Result for verification of a COSE_Mac0.
+#[derive(Debug)]
+pub enum VerificationResult {
+    Success,
+    Failure(String),
+    Error(Error),
+}
+
+impl VerificationResult {
+    /// Result of verification.
+    ///
+    /// False implies the signature is inauthentic or the verification algorithm encountered an
+    /// error.
+    pub fn success(&self) -> bool {
+        matches!(self, VerificationResult::Success)
+    }
+
+    /// Translate to a std::result::Result.
+    ///
+    /// Converts failure reasons and errors into a String.
+    pub fn to_result(self) -> Result<(), String> {
+        match self {
+            VerificationResult::Success => Ok(()),
+            VerificationResult::Failure(reason) => Err(reason),
+            VerificationResult::Error(e) => Err(format!("{}", e)),
+        }
+    }
+
+    /// Retrieve the error if the verification algorithm encountered an error.
+    pub fn to_error(self) -> Option<Error> {
+        match self {
+            VerificationResult::Error(e) => Some(e),
+            _ => None,
+        }
+    }
+}
+
+impl PreparedCoseMac0 {
+    pub fn new(
+        builder: coset::CoseMac0Builder,
+        detached_payload: Option<Vec<u8>>,
```

Comment:
You never consume `detached_payload`, you should use `&[u8]` instead of `Vec<u8>`

Reply:
changed